### PR TITLE
[BUGFIX] Serialization logic - additional fixes.

### LINKED ISSUE #1392: [BUG] Suite display not rendered correct in IFrame
**Describe the bug**
When displaying suite result in IFrame (for example in colab) the style is not rendered correctly (very narrow and tables are ugly)

image

**To Reproduce**
Run in colab:
```
from deepchecks.tabular.datasets.classification import iris
from deepchecks.tabular.suites import single_dataset_integrity

train, test = iris.load_data()
single_dataset_integrity().run(train)
```

### PR DESCRIPTION
**Pull request by yromanyshyn**
resolves #1392 but contains much more modifications

List of changes:

== 1 == 

added a new type of serialization protocol, called `IPythonSerializer`.
Serializers that implement this kind of protocol must return a list of objects that implements `IPython.core.formatters` traits.
For more info please refer to the IPython [documentation](https://ipython.readthedocs.io/en/stable/api/generated/IPython.core.formatters.html#IPython.core.formatters.MimeBundleFormatter)

We need this type of protocol in order to make sure that the check/suite result display works properly in different environments like JupyterLab/Classical Notebooks, especially when widgets are not enabled or not supported.

== 2 ==

added `display_check`, `show`, `save_as_html` methods to the `CheckFailure` type.

== 3 ==

now, 'Widgets UI' is also used in the colab env.

### CODE REVIEW

File: deepchecks/core/serialization/check_result/wandb.py [hunk 1]
```diff
@@ -32,12 +32,11 @@
...
-try:
-    if t.TYPE_CHECKING:
-        from wandb.sdk.data_types.base_types.wb_value import \
-            WBValue  # pylint: disable=unused-import
-except ImportError:
-    pass
+
+if t.TYPE_CHECKING:
+    from wandb.sdk.data_types.base_types.wb_value import \
+        WBValue  # pylint: disable=unused-import
```
Comment (JKL98ISR) [inline deepchecks/core/serialization/check_result/wandb.py:38]: you should keep the try catch because it was only added to the newest wandb version and it is only for type checking and we shouldn't make users update it

File: deepchecks/core/check_result.py [hunk 2]
```diff
@@ -146,40 +159,71 @@ def __init__(
...
+        return {'name': check_name, 'params': parameters, 'header': header,
+                'summary': get_docs_summary(self.check, with_doc_link=with_doc_link)}
 
-        Parameters
-        ----------
-        unique_id : str
-            The unique id given by the suite that displays the check.
-        show_additional_outputs : bool
-            Boolean that controls if to show additional outputs.
+    def get_header(self) -> str:
```
Comment (JKL98ISR) [inline deepchecks/core/check_result.py:169]: remove methods that are already in BaseCheckResult

File: deepchecks/core/serialization/suite_result/wandb.py [hunk 3]
```diff
@@ -21,11 +21,13 @@
...
 
 try:
-    if t.TYPE_CHECKING:
-        from wandb.sdk.data_types.base_types.wb_value import \
-            WBValue  # pylint: disable=unused-import
-except ImportError:
-    pass
+    from wandb.sdk.data_types.base_types.wb_value import \
+        WBValue  # pylint: disable=unused-import
+except ImportError as error:
```
Comment (JKL98ISR) [inline deepchecks/core/serialization/suite_result/wandb.py:26]: see other wandb comment

**Review by JKL98ISR — CHANGES_REQUESTED** [review]

File: deepchecks/core/serialization/common.py [hunk 4]
```diff
@@ -336,3 +338,29 @@ def concatv_images(images, gap=10):
...
 
         return dst
+
+
+T = t.TypeVar('T')
+DeepIterable = t.Iterable[t.Union[T, 'DeepIterable[T]']]
+
+
+def flatten(l: DeepIterable[T]) -> t.Iterable[T]:
+    """Flatten nested terables."""
```
Comment (JKL98ISR) [inline deepchecks/core/serialization/common.py:348]:
```suggestion
    """Flatten nested iterables."""
```

File: deepchecks/core/check_result.py (hunk 2 above)
Comment (yromanyshyn) [reply deepchecks/core/check_result.py]: removed

File: deepchecks/core/serialization/check_result/wandb.py (hunk 1 above)
Comment (yromanyshyn) [reply deepchecks/core/serialization/check_result/wandb.py:38]: from python docs

>typing.TYPE_CHECKING - A special constant that is assumed to be True by 3rd party static type checkers. It is False at runtime.

I am not sure if that try-except is needed here/ in general it will never raise an error, and static type checker probably will mark it as unresolved import/

File: deepchecks/core/serialization/common.py (hunk 4 above)
Comment (yromanyshyn) [reply deepchecks/core/serialization/common.py]: fixed

File: deepchecks/core/serialization/suite_result/wandb.py (hunk 3 above)
Comment (yromanyshyn) [reply deepchecks/core/serialization/suite_result/wandb.py]: replaced with `if t.TYPE_CHECKING` statement

File: deepchecks/core/serialization/check_result/wandb.py (hunk 1 above)
Comment (JKL98ISR) [reply deepchecks/core/serialization/check_result/wandb.py:38]: oh ok

**Review by JKL98ISR — APPROVED** [review]

File: deepchecks/core/suite.py [hunk 5]
```diff
@@ -143,24 +173,27 @@ def save_as_html(
...
+        """
         if file is None:
             file = 'output.html'
         if isinstance(file, str):
             file = create_new_file_name(file)
 
         if as_widget is True:
             widget_to_html(
-                widget=self.to_widget(unique_id=output_id),
+                widget=self.to_widget(unique_id=unique_id or get_random_string(n=25)),
```
Comment (JKL98ISR) [inline deepchecks/core/suite.py:189]: why the change?

**Review by JKL98ISR — CHANGES_REQUESTED** [review]

File: deepchecks/core/suite.py (hunk 5 above)
Comment (yromanyshyn) [reply deepchecks/core/suite.py:189]: previously `output_id` was declared above `if file is None:` now I just replaced it with one statement/

it means that if a user did not pass a `unique_id` we need to generate it by ourselves in order to make sure that navigation links will be rendered

**Review by JKL98ISR — APPROVED** [review]

**Review by JKL98ISR — APPROVED** [review]
check.show doesn't work in jupyter lab

**Review by JKL98ISR — CHANGES_REQUESTED** [review]
check.show doesn't work in jupyter lab

**Review by JKL98ISR — APPROVED** [review]
did a mistake while checking :)
